### PR TITLE
[AIRFLOW-489] Add Further API Functionality

### PR DESCRIPTION
**Pull request by robin-miller-ow**
Dear Airflow Maintainers,

Please accept this PR that contributes towards the following issues:
- https://issues.apache.org/jira/browse/AIRFLOW-489 (I don't believe this is sufficient to close this issue)

Testing Done:
- Unittests have been added for each of the new API calls.

### CODE REVIEW

**Comment by bolkedebruin** [comment]
Hey @robin-miller-ow thanks for extending on the APi functionality. Couple of things

- Please split your PR per functionality you are adding
- Also make sure if you update/add endpoints to make them available in api/clients/
- execution_date should not be passed in the URL but as part of the POST request
- this also removes the need for trigger_dag_for_date it can just be handled in the original function.

**Comment by robin-miller-ow** [comment]
Hi @bolkedebruin, thanks for your feedback. I'll get working on updating this accordingly.

I wasn't sure if a new function was needed in api/clients if the functionality was already just a single call for a function of a model class, but I can easily add wrappers.

I'll prepare the first change in this PR and then create new ones for the others.

**Comment by robin-miller-ow** [comment]
Hi @bolkedebruin, I have updated this branch to contain only the first change (allowing the passing of an execution_date to trigger_dag). To avoid merge complications, I have set up the other changes to each follow on from each other, so I'll open a PR for the next once this is merged in and then one after that.

File: airflow/www/api/experimental/endpoints.py [hunk 1]
```diff
@@ -44,16 +49,34 @@ def trigger_dag(dag_id):
...
+            response = jsonify({'error': error_message})
+            response.status_code = 400
+
+            return response
+
     try:
-        dr = trigger.trigger_dag(dag_id, run_id, conf)
+        dr = trigger.trigger_dag(dag_id, run_id, conf, execution_date)
     except AirflowException as err:
-        logging.error(err)
```
Comment (bolkedebruin) [inline airflow/www/api/experimental/endpoints.py:50]: Why are you removing the error logging on the server side?

Comment (robin-miller-ow) [reply airflow/www/api/experimental/endpoints.py:50]: Ah, that's my mistake. In the other endpoints I'm adding, I put the logging in the method in api/common so if it's called from elsewhere and isn't logged there, it still gets logged. I then removed the logging from here so it wasn't logged twice and did the same thing to this one even though I hadn't added the logging in trigger_dag.py. Given that I'll be updating this either way, would you prefer it here or in trigger_dag.py?

Comment (bolkedebruin) [reply airflow/www/api/experimental/endpoints.py:50]: Logging of exceptions should only be at the top-level, thus in this case in here (endpoint).

**Comment by bolkedebruin** [comment]
@robin-miller-ow any update? In case you want this to be part of 1.8 it needs to pass quickly.

**Comment by robin-miller-ow** [comment]
@bolkedebruin sorry for the silence, I've been really busy this week. I've just pushed a version of the branch with that log call back in.

**Comment by bolkedebruin** [comment]
@robin-miller-ow Thanks. Please don't forget to update "cli.py" to allow the execution_date to be set as well from the command line (API and cli should be in sync as the cli can call the API)

**Comment by robin-miller-ow** [comment]
@bolkedebruin I believe I've added wired it up in cli.py now. I was having a look at the tests for that to see if I could expand it. I think the only test is at line 1246 in tests/core.py and I don't think I fully understand what it's doing. Is it worth trying to improve it at this time?

**Comment by robin-miller-ow** [comment]
@bolkedebruin I thought I'd found what I'd missed on this, but the unit-test is still failing, now with the exceptionally unhelpful error message "SystemExit: 2". I'll get back on this on Monday morning, but would very much appreciate any thoughts you have on what I've broken.

**Comment by bolkedebruin** [comment]
File "/opt/python/2.7.12/lib/python2.7/argparse.py", line 1950, in _parse_known_args
    self.error(_('too few arguments'))


I did not look at your code, but I am assuming your are requiring "execution_date" to be set instead of making it optional.

**Comment by robin-miller-ow** [comment]
@bolkedebruin you were quite right. I was trying to use the existing execution_date argument set up for use with other airflow [command] calls, but this is set as a required positional parameter for these calls. As such I've had to call the parameter "exec_date" instead. When I tried to set it up with the verbose flag: 
"--execution_date" I still got an error, so I tried the verbose flag "--exec_date" instead. This then worked, but showed up that I'd missed the pass through calls in the api/client folder, which I've now fixed. Tests are passing. LandscapeIO is unhappy, but seems to have picked up issues in files that I've not touched and none new in the files that I have touched. As such, I'm not intending to fix the issues it says I've introduced.
Unless you have any further comments, I believe this is ready to merge.

File: airflow/api/client/json_client.py [hunk 2]
```diff
@@ -29,6 +29,7 @@ def trigger_dag(self, dag_id, run_id=None, conf=None):
                              json={
                                  "run_id": run_id,
                                  "conf": conf,
+                                 "exec_date": execution_date,
```
Comment (bolkedebruin) [inline airflow/api/client/json_client.py]: please spell it out: "execution_date"

File: airflow/www/api/experimental/endpoints.py [hunk 3]
```diff
@@ -44,16 +49,35 @@ def trigger_dag(dag_id):
     if 'conf' in data:
         conf = data['conf']
 
+    execution_date = None
+    if 'exec_date' in data:
```
Comment (bolkedebruin) [inline airflow/www/api/experimental/endpoints.py]: please spell it out for json. "execution_date"

File: airflow/www/api/experimental/endpoints.py [hunk 4]
```diff
@@ -44,16 +49,35 @@ def trigger_dag(dag_id):
...
+        execution_date = data['exec_date']
+
+        # Convert string datetime into actual datetime
+        try:
+            execution_date = datetime.strptime(execution_date,
+                                               '%Y-%m-%dT%H:%M:%S')
```
Comment (bolkedebruin) [inline airflow/www/api/experimental/endpoints.py:59]: I don't think this is ISO 8601 compliant: https://en.wikipedia.org/wiki/ISO_8601

UPDATE: seems I am wrong, nevermind.

File: airflow/models.py [hunk 5]
```diff
@@ -2829,6 +2829,24 @@ def get_active_runs(self, session=None):
 
         return active_dates
 
+    @provide_session
```
Comment (bolkedebruin) [inline airflow/models.py]: This is duplicate code. Probably merged in a recent PR.

**Comment by bolkedebruin** [comment]
Some minor nits. Please correct and I'll merge.

**Pull request closed without merging**